_posts/2020-12-10-webpack-vs-sprockets.markdown

```diff
@@ -0,0 +1,313 @@
+
+```ruby
+# config/initializers/assets.rb
+
+Rails.application.config.assets.precompile += %w( admin.js admin.css )
+```
+
+To do the same with Webpack, you don't need to change a configuration. All the files at `app/javascript/packs` (AKA "the entry points") will be created (AKA "emitted"). You can change where your packs are located too in `config/webpacker.yml`:
+
+```
+  source_entry_path: packs
+```
+
+By default, you have an `application.js` file there, but you can add an `admin.js` one too for example:
+
+```js
+// app/javascript/packs/admin.js
+
+// here you can add all your code or require other files
+```
+
+Now, when Webpack compiles your assets, it will emit `application.js` and `admin.js`.
+
+You can also create a `.css` (or `.scss` if you prefer) file to be emitted:
+
+```sass
+// app/javascript/packs/admin.scss
+
+@import some_sass_module // you can use SASS imports
+```
+
+And now Webpack will also process, compile and emit an `admin.css` file.
+> There's a caveat when using a CSS pack, I'll comment that later when I talk about images
+
+> Note that **ALL** the files under `/packs` will be emitted. You don't want to put all your source files there, but only the ones you are going to access directly! All your source files should be at the parent folder or at a sibling folder.
+
+## Node Modules
+
+Rails uses [YARN](https://yarnpkg.com/) by default to handle Node packages. All packages are downloaded in a `node_modules` folder in the root of your project.
+
+> Remember to add that folder to the `.gitignore` file, you don't want to push all these files into your repo
+
+These packages can be used by both Sprockets and Webpack, so you can use the same package to provide some CSS for your stylesheets using Sprockets and some JavaScript for your scripts using Webpack.
+
+If you want Sprockets to look for files inside that folder, you need to add that to the assets paths list:
+
+```ruby
+# config/initializers/assets.rb
+
+Rails.application.config.assets.paths << Rails.root.join('node_modules')
+```
+
+Webpack will look for files there by default, so no change needed, but if you want to tell Webpack to check other folders too you can add paths to the `additional_paths` configuration at `config/webpacker.yml`.
+
+Now, when you `require` or `import` in a JavaScript file, `@import` on an SCSS file or `// require` on a CSS file, the compilers will look for a folder with the name we used there on any asset path. When we see a line like this:
+
+```js
+require('@rails/ujs').start()
+```
+
+It is looking for a module at `node_modules/@rails/ujs`, and it checks the `package.json` file inside that folder to know what to load. You can also reference specific files instead of a module, for example, the `@rails/ujs` package's `package.json` file has this line:
```

Comment:
```suggestion
It is looking for a module in `node_modules/@rails/ujs`, and it checks the `package.json` file inside that folder to know what to load. You can also reference specific files instead of a module, for example, the `@rails/ujs` package's `package.json` file has this line:
```